src/beamlime/applications/base.py

```diff
@@ -0,0 +1,176 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+import asyncio
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, Coroutine, Generator, List, Optional
+
+from beamlime.logging import BeamlimeLogger
+
+from ..logging.mixins import LogMixin
+
+
+class BaseDaemon(LogMixin, ABC):
+    """Base class for daemons.
+
+    Daemons are long-running processes that handle events/messages.
+    It is expected that the ``run`` method is called in an event loop.
+    Daemons handle events/messages within the ``run`` method.
+    """
+
+    logger: BeamlimeLogger
+
+    def data_pipe_monitor(
+        self,
+        pipe: List[Any],
+        timeout: float = 5,
+        interval: float = 1 / 14,
+        preferred_size: int = 1,
+        target_size: int = 1,
+    ):
+        from beamlime.core.schedulers import async_retry
+
+        @async_retry(
+            TimeoutError, max_trials=int(timeout / interval), interval=interval
+        )
+        async def wait_for_preferred_size() -> None:
+            if len(pipe) < preferred_size:
+                raise TimeoutError
+
+        async def is_pipe_filled() -> bool:
+            try:
+                await wait_for_preferred_size()
+            except TimeoutError:
+                await asyncio.sleep(0)  # Let other handlers use the event loop.
+            return len(pipe) >= target_size
+
+        return is_pipe_filled
+
+    @abstractmethod
+    async def run(self):
+        ...
+
+
+@dataclass
+class BeamlimeMessage:
+    """A message object that can be sent through a message router."""
+
+    sender: Any
+    receiver: Any
+    content: Any
+
+
+class MessageRouter(BaseDaemon):
+    """A message router that routes messages to handlers."""
+
+    message_pipe: List[BeamlimeMessage]
+
+    def __init__(self, message_pipe: List[BeamlimeMessage]):
+        self.handlers: dict[type, List[Callable[[BeamlimeMessage], Any]]] = dict()
+        self.awaitable_handlers: dict[
+            type, List[Callable[[BeamlimeMessage], Awaitable[Any]]]
+        ] = dict()
+        self.message_pipe = message_pipe
+        self._break_routing_loop = False
+
+        class StopRouting(BeamlimeMessage):
```

Comment:
The original one was using ``while-loop`` that breaks based on the ``StopWatch``
because it was mainly for benchmarking tests.
``MessageRouter`` was inserted to replace that.
I didn't have better idea than this... if it's too confusing I can remove it and go back to the original way.